### ISSUE TITLE
Playwright --watch so not all apps have to restart when code/tests change

### ISSUE THREAD

**Issue by Langstra** [TriPSs/nx-extend#81]
Currently when you run the tests, it will start the application or applications if you have the dependencies. Then it will run the tests and kill playwright. The process of starting the apps and playwright takes some time. This also makes that rerunning tests takes quite some time, even though the change to a test is very small.

It would be very useful if the applications could continue to run after the playwright test run had finished. Then it should watch for code change to the tests and rerun the tests if needed.

**Comment by TriPSs** [comment]
Do you use `e2e-runner`? Cause there you have the option `reuseExistingServer`, then you could start the needed servers yourself and it will only start playwright.

As far as I know playwright it self does not have a watch option but I'm not quite sure, havent used it enough yet.

**Comment by Langstra** [comment]
Yes I do use the `e2e-runner` with the `reuseExistingServer`. However, while the servers are starting playwright itself is also starting. Then playwright detects that the servers are not running and is going to start the servers. This leads to the problem that two instances of the servers are going to start.

**Comment by TriPSs** [comment]
Could you elaborate a bit on the `while the servers are starting playwright itself is also starting`, cause usually what I do is start all the servers in different terminals and after they are all started than start playwright?

Beside that, I do think adding something like `--watch` would be nice addition to Playwright.

This is something Playwright itself does not do ([#7035](https://github.com/microsoft/playwright/issues/7035)) so we would have to build it ourself.

**Comment by Langstra** [comment]
Ah yeah, w

> Could you elaborate a bit on the `while the servers are starting playwright itself is also starting`, cause usually what I do is start all the servers in different terminals and after they are all started than start playwright?

Ah yeah, what I do is start the servers at the same time as playwright. I do this in my CI. In the pipeline I cannot check if the servers have already started before I start playwright. Maybe we could also add an option that `doNotStart` of just not to start the services when `reuseExistingServer` is set to `true`

Another great addition would be to disable TLS verification, but maybe I should open a new issue for that.

**Comment by bard** [comment]
> Ah yeah, what I do is start the servers at the same time as playwright. I do this in my CI. In the pipeline I cannot check if the servers have already started before I start playwright.

Maybe https://www.npmjs.com/package/start-server-and-test helps?

**Comment by Langstra** [comment]
Yeah, I'm currently using that, but I was hoping to ditch that extra package ;)

**Comment by jfgreffier** [comment]
> Maybe https://www.npmjs.com/package/start-server-and-test helps?

Let Playwright launch the server and wait for connectivity is maybe easier ? At least it can avoid installing an extra package.

```ts
// playwright.config.ts
import type { PlaywrightTestConfig } from '@playwright/test';
const config: PlaywrightTestConfig = {
  webServer: {
    command: 'npm run start',
    url: 'http://localhost:3000/app/',
    timeout: 120 * 1000,
    reuseExistingServer: !process.env.CI,
  },
  use: {
    baseURL: 'http://localhost:3000/app/',
  },
};
export default config;
```


https://playwright.dev/docs/test-advanced#launching-a-development-web-server-during-the-tests

**Comment by jfgreffier** [comment]
> This is something Playwright itself does not do ([#7035](https://github.com/microsoft/playwright/issues/7035)) so we would have to build it ourself.

If someone could contribute watch mode to Playwright, it would be even better ^^

**Comment by Langstra** [comment]
> Let Playwright launch the server and wait for connectivity is maybe easier ? At least it can avoid installing an extra package.

Problem here is that playwright also kills the server after finishing the tests. This makes rerunning the tests after small changes a time-consuming process. Our Angular applications, and api, take some time to startup. So we'd rather start them once and then run the tests a few times.

**Comment by jfgreffier** [comment]
> > Let Playwright launch the server and wait for connectivity is maybe easier ? At least it can avoid installing an extra package.
> 
> Problem here is that playwright also kills the server after finishing the tests. This makes rerunning the tests after small changes a time-consuming process. Our Angular applications, and api, take some time to startup. So we'd rather start them once and then run the tests a few times.

Ah yes, it totally makes sense.

**Comment by TriPSs** [comment]
> > Let Playwright launch the server and wait for connectivity is maybe easier ? At least it can avoid installing an extra package.
> 
> Problem here is that playwright also kills the server after finishing the tests. This makes rerunning the tests after small changes a time-consuming process. Our Angular applications, and api, take some time to startup. So we'd rather start them once and then run the tests a few times.

Was adding new tests recently and was noticing the same. One option could be to define a `project.e2e.configurations.<your name>` that has the `reuseExistingServer` set to `true` and run using that config, that way you can start everything once just run all the tests.

That is something that could make working with Playwright a bit easier until they release something like https://github.com/microsoft/playwright/issues/7035

**Comment by TriPSs** [comment]
Playwright@[v1.32.0](https://github.com/microsoft/playwright/releases/tag/v1.32.0) latest release now has a ui with watch mode option 🎉
